test/test_list.cpp

```diff
@@ -0,0 +1,716 @@
+        CHECK(!avg->is_null());
+        CHECK_EQUAL(ret_count, 1);
+        CHECK_EQUAL(avg->template get<ColumnAverageType<T>>(), ColumnAverageType<T>(1));
+    }
+
+    t->rollback();
+}
+
+TEST(List_AggOps)
+{
+    SHARED_GROUP_TEST_PATH(path);
+
+    std::unique_ptr<Replication> hist(make_in_realm_history());
+    DBRef sg = DB::create(*hist, path, DBOptions(crypt_key()));
+
+    test_lists_numeric_agg<int64_t>(test_context, sg, type_Int);
+    test_lists_numeric_agg<float>(test_context, sg, type_Float);
+    test_lists_numeric_agg<double>(test_context, sg, type_Double);
+    test_lists_numeric_agg<Decimal128>(test_context, sg, type_Decimal);
+
+    test_lists_numeric_agg<Optional<int64_t>>(test_context, sg, type_Int, Optional<int64_t>{}, true);
+    test_lists_numeric_agg<float>(test_context, sg, type_Float, realm::null::get_null_float<float>(), true);
+    test_lists_numeric_agg<double>(test_context, sg, type_Double, realm::null::get_null_float<double>(), true);
+    test_lists_numeric_agg<Decimal128>(test_context, sg, type_Decimal, Decimal128(realm::null()), true);
+}
+
+TEST(List_NestedListColumns)
+{
+    SHARED_GROUP_TEST_PATH(path);
+    DBRef db = DB::create(make_in_realm_history(), path);
+    auto tr = db->start_write();
+    auto table = tr->add_table("table");
+    auto int_col = table->add_column(type_Int, "int");
+    auto int_list_col = table->add_column(type_Int, "int_list", false, {CollectionType::List});
+    auto list_col1 =
+        table->add_column(type_Int, "int_list_list", false, {CollectionType::List, CollectionType::List});
+    auto list_col2 = table->add_column(type_Int, "int_dict_list_list", false,
+                                       {CollectionType::Dictionary, CollectionType::List, CollectionType::List});
+
+    tr->commit_and_continue_as_read();
+    CHECK_EQUAL(table->get_nesting_levels(int_col), 0);
+    CHECK(!int_col.is_list());
+    CHECK_EQUAL(table->get_nesting_levels(int_list_col), 0);
+    CHECK(int_list_col.is_list());
+    CHECK_EQUAL(table->get_nesting_levels(list_col1), 1);
+    CHECK_EQUAL(table->get_nesting_levels(list_col2), 2);
+}
+
+TEST(List_NestedList)
+{
+    SHARED_GROUP_TEST_PATH(path);
+    DBRef db = DB::create(make_in_realm_history(), path);
+    auto tr = db->start_write();
+    auto table = tr->add_table("table");
+    auto list_col1 =
+        table->add_column(type_Int, "int_list_list", false, {CollectionType::List, CollectionType::List});
+    auto list_col2 = table->add_column(type_Int, "int_dict_list_list", false,
+                                       {CollectionType::Dictionary, CollectionType::List, CollectionType::List});
+    CHECK_EQUAL(table->get_nesting_levels(list_col1), 1);
+    CHECK_EQUAL(table->get_nesting_levels(list_col2), 2);
+    Obj obj = table->create_object();
+
+    auto list = obj.get_collection_list(list_col1);
+    CHECK(list->is_empty());
+    auto collection = list->insert_collection(0);
+    dynamic_cast<Lst<Int>*>(collection.get())->add(5);
```

Comment:
SDKs will probably want a more typed API than using dynamic casts. At least that has been the traditional approach.

Reply:
Yes, lets see. Anyway that would be implemented in the ObjectStore interface.

Reply:
Anyway - I added a couple of functions. Will actually make some testing easier to write.